kernel/src/sched/sched_class/mod.rs

```diff
@@ -41,6 +41,8 @@ use super::{
 };
 use crate::thread::{AsThread, Thread};
 
+type SchedEntity = (Arc<Task>, Arc<Thread>);
```

Comment:
This is ugly. It saves some calls to `Arc::upgrade` in `Thread::task`, but adds memory footprint overhead in every place that stores `SchedEntity`. So this may not bring net positive result.

I think you should consider maintaining an `Arc<Thread>` in `PerCpuClassRqSet::current`.

```rust
struct PerCpuClassRqSet {
    ...
    current: Option<(Arc<Task>, Arc<Thread> /* this is new */, CurrentRuntime)>,
}
```

This only adds memory storage cost in only one place. In exchange, all the calls to `Task::as_thread` in the critical paths of `pick_next_current` and `update_current` can be avoided.

The `pick_next_current` method has another clone that is unnecessary.

```rust
    fn pick_next_current(&mut self) -> Option<&Arc<Task>> {
        self.pick_next_thread().and_then(|next| {
            let next_task = next.task();
            if let Some((old_task, _)) = self
                .current
                // This clone is unnecessary because next_thread still exists, just inside current.0
                .replace((next_task.clone(), CurrentRuntime::new()))
            {
                if Arc::ptr_eq(&old_task, &next_task) {
                    return None;
                }
                let old = old_task.as_thread().unwrap();
                self.enqueue_thread(old, None);
            }
            self.current.as_ref().map(|(task, _)| task)
        })
    }
```

Reply:
On one hand, an `Arc<Task>` owns an `Arc<Thread>` but not in reverse, so storing `Arc<Thread>`s alone may drop the underlying task. On the other hand, every piece of information needs to be retrieved in `Thread`s but not tasks, so storing `Arc<Task>`s alone dramatically downgrades the performance. Hence, preserving both might be a reasonable choice before the relationship between tasks & threads changes.